src/main/java/org/littleshoot/proxy/impl/ProxyConnection.java

```diff
@@ -742,6 +743,46 @@ public void channelRead(ChannelHandlerContext ctx, Object msg)
         protected abstract void responseRead(HttpResponse httpResponse);
     }
 
+    @Sharable
+    protected class InboundGlobalStateHandler extends
+        ChannelInboundHandlerAdapter {
+
+        private final ProxyConnection<HttpRequest> clientToProxyConnection;
+
+        InboundGlobalStateHandler(ProxyConnection<HttpRequest> clientToProxyConnection) {
+            this.clientToProxyConnection = clientToProxyConnection;
+        }
+
+        @Override
+        public void channelRead(ChannelHandlerContext ctx, Object msg)
+            throws Exception {
+            proxyServer.getGlobalStateHandler().restoreFromChannel(clientToProxyConnection.channel);
+            super.channelRead(ctx, msg);
+            proxyServer.getGlobalStateHandler().clear();
+        }
+    }
```

Comment:
`super.channelRead(ctx, msg);` takes the complete pipeline. Inside it does data processing and then calls `fireChannelRead` which takes a next handler in the chain and calls the same `channelRead` on it after which the process repeats until the last handler is reached. After that the method `super.channelRead(ctx, msg);` is returned and re clear the context.